src/App.js

```diff
@@ -1,53 +1,145 @@
+
+function mapProfileData(rawData){
+  return rawData.map(({name, email, dob, picture}) => ({
+    name: name.first + ' ' + name.last,
+    email,
+    age: dob.age,
+    imgUrl: picture.large
+  }))
+}
+let remainingCardsNum;
+const REMAINING_CARD_THRESHOLD = 2;
 
 function App() {
   const [profiles, setProfiles] = useState([]);
   const [currentProfileIndex, setProfileIndex] = useState(0);
+  const [viewedProfiles, setViewedProfiles] = useState([]);
+  const [shouldPrefetchProfile, setShouldPrefetchProfile] = useState(false);
+  const [isLoading, setLoading] = useState(false)
 
   useEffect(() => {
-    setProfiles(profileData)
+    (async function getData (){
+      setLoading(true)
+      const {results: rawData} = await getProfilesData();
+      const profileData = mapProfileData(rawData)
+      setProfiles(profileData)
+      setLoading(false)
+      remainingCardsNum = profileData.length;
+    })()  
   }, [])
 
+  useEffect(() => {
+    if(shouldPrefetchProfile){
+      (async function getData (){  
+        notification['success']({
+          message: 'Prefetch 5 more cards',
+          description:
+            `There is only 1 left, time for getting more!` ,
+          duration: 1
+          });    
+        const {results: rawData} = await getProfilesData();
+        const profileData = mapProfileData(rawData)   
+        remainingCardsNum = profileData.length + remainingCardsNum;  
+        setProfiles(profiles.concat(profileData))
+        setLoading(false)
+        setShouldPrefetchProfile(false)         
+      })()
+    } 
+  }, [shouldPrefetchProfile])
+
+  useEffect(() => {
+    const isCurrentCardExist = currentProfileIndex <= profiles.length;
+    isCurrentCardExist && setLoading(false)
+  }, [isLoading])
+
   const debouncedSwipe = _.debounce(function handleSwipe(type){
-    const isLastCard = profiles.length - 1 === currentProfileIndex;
-    if(type === 'like'){
-      message.success('What an awesome profile!', 0.3)
-    } else {
-      message.error('Not my type', 0.3)
+    processCurrentCardAction();
+    moveToNextCard();
+
+    function processCurrentCardAction(){
+      const changedViewedProfiles = viewedProfiles.concat(
+        {...profiles[currentProfileIndex], liked: type === 'like' ? true: false}
+      )
+      setViewedProfiles(changedViewedProfiles); 
+      type === 'like' ?
+        message.success('What an awesome profile!', 0.3)
+        : message.error('Not my type', 0.3);
     }
-    setProfileIndex(
-      !isLastCard
-      ? currentProfileIndex + 1
-      : 0
-    );
-  }, 500)
+    
+    function moveToNextCard(){
+      setProfileIndex(currentProfileIndex + 1);
+      const isNextCardExist = profiles[currentProfileIndex + 1];
+      if(!isNextCardExist) {
+        setLoading(true) 
+        notification['warning']({
+          message: 'Oops! Seem like internet connection is slow',
+          duration: 1
+        }); 
+      }
+      --remainingCardsNum
```

Comment:
curious why the preference `--remainingCardsNum` instead of just `remainingCardsNum -= 1`

Reply:
This shall be removed in [this commit]("https://github.com/SophiePham92/react-tinder-cards/pull/5/commits/f8fb4a19ff022bc89d50cf307a77b3260ed00584")